monitor/health/http.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package health
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/ChainSafe/ChainBridge/core"
+	msg "github.com/ChainSafe/ChainBridge/message"
+	log "github.com/ChainSafe/log15"
+)
+
+type httpMetricServer struct {
+	port         int
+	timeDelay    int
+	core         *core.Core
+	blockheights map[msg.ChainId]blockHeightInfo
+}
+
+type blockHeightInfo struct {
+	height      *big.Int
+	lastUpdated time.Time
+}
+
+type httpMetricOptions struct {
+	port      int
+	timeDelay int
+	core      *core.Core
+}
+
+type httpResponse struct {
+	Data  string
+	Error string
+}
+
+func newhttpMetricServer(opts httpMetricOptions) *httpMetricServer {
+	return &httpMetricServer{
+		port:         opts.port,
+		core:         opts.core,
+		timeDelay:    opts.timeDelay,
+		blockheights: make(map[msg.ChainId]blockHeightInfo),
+	}
+}
+
+// Start starts the http metrics server
+func (s httpMetricServer) Start() {
+	log.Info("Metrics server started", "port", s.port)
+
+	// Setup routes
+	http.HandleFunc("/health", s.healthStatus)
+
+	// Start http server
+	err := http.ListenAndServe(":"+strconv.Itoa(s.port), nil)
+
+	if err == http.ErrServerClosed {
+		log.Info("Server is shutting down", err)
+	} else {
+		log.Error("Shutting down, server error: ", err)
+	}
+}
+
+// healthStatus is a catch-all update that grabs the latest updates on the running chains
+// It assumes that the configuration was set correctly, therefore the relevant chains are
+// only those that are in the core.Core registry.
+func (s httpMetricServer) healthStatus(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	// Grab all chains
+	chains := s.core.GetChains()
+	requestTime := time.Now()
+
+	// Iterate through their block heads and update the cache accordingly
+	for _, chain := range chains {
+		latestHeight := chain.GetLatestBlock()
+
+		// Get old blockheight
+		if prevHeight, ok := s.blockheights[chain.Id()]; ok {
+			// Note The listener performing the polling should already be accounting for the block delay
+			// TODO Account for timestamps
+			timeDiff := requestTime.Sub(prevHeight.lastUpdated)
+			if latestHeight.Cmp(prevHeight.height) >= 0 && int(timeDiff.Seconds()) < s.timeDelay {
+				s.blockheights[chain.Id()] = blockHeightInfo{
+					height:      latestHeight,
+					lastUpdated: requestTime,
+				}
+				return
```

Comment:
Why return inside these if clauses?

Reply:
I removed this one, the other ones are there because the writer doesn't cause it to exit, so what ends up happening is an inner `if/else` may trigger an error, and right after line 129 will trigger